File: frontend/src/types/cashflow.ts
export interface MonthlyCashflow {
  cf_id?: number;
  project_id: number;
  month_number: number;
  calendar_date?: string;
  construction_cost: number;
  contingency: number;
  horizontal: number;
  soft_costs: number;
  gc_fee: number;
  land: number;
  other: number;
  total_development_cost: number;
  cumulative_development_cost: number;
  equity_draw: number;
  senior_draw: number;
  senior_balance: number;
  senior_interest: number;
  mezz_draw: number;
  mezz_balance: number;
  mezz_interest: number;
  land_loan_draw: number;
  land_loan_balance: number;
  gross_sales: number;
  cost_of_sale: number;
  net_sales: number;
  loan_payoff_from_sales: number;
  free_cashflow: number;
  cumulative_cashflow: number;
}

export interface WaterfallResult {
  gp_return_of_capital: number;
  lp_return_of_capital: number;
  gp_preferred_return: number;
  lp_preferred_return: number;
  gp_promote_by_tier: Record<string, number>;
  lp_promote_by_tier: Record<string, number>;
  total_gp: number;
  total_lp: number;
  gp_multiple: number;
  lp_multiple: number;
}

export interface ProjectMetrics {
  total_development_cost: number;
  total_with_financing: number;
  total_revenue: number;
  total_cost_of_sale: number;
  senior_debt_committed: number;
  total_equity: number;
  gp_equity: number;
  lp_equity: number;
  senior_interest: number;
  project_profit: number;
  profit_margin: number;
  levered_xirr?: number | null;
  unlevered_xirr?: number | null;
  equity_multiple?: number | null;
  gp_xirr?: number | null;
  lp_xirr?: number | null;
  gp_multiple?: number | null;
  lp_multiple?: number | null;
  ltc_ratio?: number | null;
  peak_equity: number;
  peak_debt_balance: number;
  total_financing_cost: number;
  waterfall?: WaterfallResult | null;
}

export interface RecalcResponse {
  cashflows: MonthlyCashflow[];
  metrics: ProjectMetrics;
}

export interface SensitivityCell {
  sale_price_psf: number;
  construction_cost_psf: number;
  project_profit: number;
  xirr?: number | null;
}

export interface SensitivityGrid {
  sale_price_psf_values: number[];
  construction_cost_psf_values: number[];
  cells: SensitivityCell[][];
}

export interface PriceSolverRequest {
  target_metric: 'profit_margin' | 'equity_multiple' | 'levered_irr';
  target_value: number;
  distribution_mode: 'uniform' | 'proportional';
}

export interface BuildingPriceSuggestion {
  building_id: number;
  building_name: string;
  suggested_psf: number;
  unit_count: number;
  total_sf: number;
}

export interface PriceSolverResponse {
  required_avg_psf: number;
  break_even_psf: number;
  target_revenue: number;
  building_suggestions: BuildingPriceSuggestion[];
}

export interface PricingAnalysis {
  break_even_psf: number;
  current_avg_psf: number;
  margin_above_break_even: number;
  psf_for_20pct_margin: number;
  psf_for_3x_multiple: number;
}

File: backend/app/schemas/cashflow.py
from pydantic import BaseModel
from typing import Optional, List, Dict
from datetime import datetime


class MonthlyCashflowResponse(BaseModel):
    cf_id: Optional[int] = None
    project_id: int
    month_number: int
    calendar_date: Optional[datetime] = None
    construction_cost: float = 0.0
    contingency: float = 0.0
    horizontal: float = 0.0
    soft_costs: float = 0.0
    gc_fee: float = 0.0
    land: float = 0.0
    other: float = 0.0
    total_development_cost: float = 0.0
    cumulative_development_cost: float = 0.0
    equity_draw: float = 0.0
    senior_draw: float = 0.0
    senior_balance: float = 0.0
    senior_interest: float = 0.0
    mezz_draw: float = 0.0
    mezz_balance: float = 0.0
    mezz_interest: float = 0.0
    land_loan_draw: float = 0.0
    land_loan_balance: float = 0.0
    gross_sales: float = 0.0
    cost_of_sale: float = 0.0
    net_sales: float = 0.0
    loan_payoff_from_sales: float = 0.0
    free_cashflow: float = 0.0
    cumulative_cashflow: float = 0.0

    class Config:
        from_attributes = True


class WaterfallTierDetail(BaseModel):
    sequence: int
    gp_amount: float
    lp_amount: float


class WaterfallResponse(BaseModel):
    gp_return_of_capital: float = 0.0
    lp_return_of_capital: float = 0.0
    gp_preferred_return: float = 0.0
    lp_preferred_return: float = 0.0
    gp_promote_by_tier: Dict[str, float] = {}
    lp_promote_by_tier: Dict[str, float] = {}
    total_gp: float = 0.0
    total_lp: float = 0.0
    gp_multiple: float = 0.0
    lp_multiple: float = 0.0


class ProjectMetricsResponse(BaseModel):
    total_development_cost: float = 0.0
    total_with_financing: float = 0.0
    total_revenue: float = 0.0
    total_cost_of_sale: float = 0.0
    senior_debt_committed: float = 0.0
    total_equity: float = 0.0
    gp_equity: float = 0.0
    lp_equity: float = 0.0
    senior_interest: float = 0.0
    project_profit: float = 0.0
    profit_margin: float = 0.0
    levered_xirr: Optional[float] = None
    unlevered_xirr: Optional[float] = None
    equity_multiple: Optional[float] = None
    gp_xirr: Optional[float] = None
    lp_xirr: Optional[float] = None
    gp_multiple: Optional[float] = None
    lp_multiple: Optional[float] = None
    ltc_ratio: Optional[float] = None
    peak_equity: float = 0.0
    peak_debt_balance: float = 0.0
    total_financing_cost: float = 0.0
    waterfall: Optional[WaterfallResponse] = None


class RecalcResponse(BaseModel):
    cashflows: List[MonthlyCashflowResponse]
    metrics: ProjectMetricsResponse


class SensitivityCell(BaseModel):
    sale_price_psf: float
    construction_cost_psf: float
    project_profit: float
    xirr: Optional[float] = None


class SensitivityGridResponse(BaseModel):
    sale_price_psf_values: List[float]
    construction_cost_psf_values: List[float]
    cells: List[List[SensitivityCell]]


class PriceSolverRequest(BaseModel):
    target_metric: str  # "profit_margin", "equity_multiple", or "levered_irr"
    target_value: float
    distribution_mode: str = "uniform"  # "uniform" or "proportional"


class BuildingPriceSuggestion(BaseModel):
    building_id: int
    building_name: str
    suggested_psf: float
    unit_count: int
    total_sf: int


class PriceSolverResponse(BaseModel):
    required_avg_psf: float
    break_even_psf: float
    target_revenue: float
    building_suggestions: List[BuildingPriceSuggestion]


class PricingAnalysisResponse(BaseModel):
    break_even_psf: float = 0.0
    current_avg_psf: float = 0.0
    margin_above_break_even: float = 0.0
    psf_for_20pct_margin: float = 0.0
    psf_for_3x_multiple: float = 0.0
